README
NAME
    Plack::Middleware::Throttle::Lite::Backend::Memcached - Memcache-driven
    storage backend for Throttle-Lite

VERSION
    version 0.02

DESCRIPTION
    This is implemetation of the storage backend for
    Plack::Middleware::Throttle::Lite. It uses memcache-server to hold
    throttling data, automatically sets expiration time for stored keys to
    save memory consumption.

SYNOPSYS
        # inside your app.psgi
        enable 'Throttle::Lite',
            backend => [
                'Memcached' => {
                    servers => [
                        'mc1.example.com:11211',
                        'mc1.example.com:11212',
                        'mc2.example.net:11210',
                    ],
                }
            ];

OPTIONS
    There are no backend-specific options. All options directly passing to
    downstream interface to memcached server. At the moment this is
    Cache::Memcached::Fast. See Cache::Memcached::Fast for available
    configuration options.

METHODS
  mc
    Returns a memcached connection handle.

  init
    See "ABSTRACT METHODS" in
    Plack::Middleware::Throttle::Lite::Backend::Abstract

  reqs_done
    See "ABSTRACT METHODS" in
    Plack::Middleware::Throttle::Lite::Backend::Abstract

  increment
    See "ABSTRACT METHODS" in
    Plack::Middleware::Throttle::Lite::Backend::Abstract

BUGS
    Please report any bugs or feature requests through the web interface at
    <https://github.com/Wu-Wu/Plack-Middleware-Throttle-Lite-Backend-Memcach
    ed/issues>

SEE ALSO
    Cache::Memcached::Fast

    Plack::Middleware::Throttle::Lite

    Plack::Middleware::Throttle::Lite::Backend::Abstract

AUTHOR
    Anton Gerasimov <chim@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Anton Gerasimov.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Plack::Middleware::Throttle::Lite::Backend::Memcached

0.02    Sun Apr 7 13:37:23 2013 +0000
    - added simple usage example

0.01    Sun Apr 7 13:12:59 2013 +0000
    - initial commit/release


Makefile.PL

use strict;
use warnings;

use 5.008009;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Memcache-driven storage backend for Throttle-Lite",
  "AUTHOR" => "Anton Gerasimov <chim\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Cache::Memcached::Fast" => 0,
    "HTTP::Message" => "6.01",
    "Plack" => "1.0001",
    "Test::Exception" => 0,
    "Test::More" => "0.98",
    "Test::Pod" => "1.22"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Plack-Middleware-Throttle-Lite-Backend-Memcached",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Plack::Middleware::Throttle::Lite::Backend::Memcached",
  "PREREQ_PM" => {
    "Cache::Memcached::Fast" => 0,
    "Carp" => 0,
    "Plack::Middleware::Throttle::Lite" => "0.02",
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.02",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Plack/Middleware/Throttle/Lite/Backend/Memcached.pm
package Plack::Middleware::Throttle::Lite::Backend::Memcached;

# ABSTRACT: Memcache-driven storage backend for Throttle-Lite

use strict;
use warnings;
use Carp ();
use parent 'Plack::Middleware::Throttle::Lite::Backend::Abstract';
use Cache::Memcached::Fast;

our $VERSION = '0.02'; # VERSION
our $AUTHORITY = 'cpan:CHIM'; # AUTHORITY

__PACKAGE__->mk_attrs(qw(mc));

sub init {
    my ($self, $args) = @_;

    my $_handle = Cache::Memcached::Fast->new($args);
    Carp::croak("Cannot get memcached handle") unless keys %{ $_handle->server_versions };

    $self->mc($_handle);
}

sub increment {
    my ($self) = @_;

    $self->mc->set($self->cache_key, 1 + $self->reqs_done, 1 + $self->expire_in);
}

sub reqs_done {
    my ($self) = @_;

    $self->mc->get($self->cache_key) || 0;
}

1; # End of Plack::Middleware::Throttle::Lite::Backend::Memcached

__END__

=pod

=head1 NAME

Plack::Middleware::Throttle::Lite::Backend::Memcached - Memcache-driven storage backend for Throttle-Lite

=head1 VERSION

version 0.02

=head1 DESCRIPTION

This is implemetation of the storage backend for B<Plack::Middleware::Throttle::Lite>. It uses memcache-server
to hold throttling data, automatically sets expiration time for stored keys to save memory consumption.

=head1 SYNOPSYS

    # inside your app.psgi
    enable 'Throttle::Lite',
        backend => [
            'Memcached' => {
                servers => [
                    'mc1.example.com:11211',
                    'mc1.example.com:11212',
                    'mc2.example.net:11210',
                ],
            }
        ];

=head1 OPTIONS

There are no backend-specific options. All options directly passing to downstream interface to memcached server.
At the moment this is B<Cache::Memcached::Fast>. See L<Cache::Memcached::Fast> for available configuration options.

=head1 METHODS

=head2 mc

Returns a memcached connection handle.

=head2 init

See L<Plack::Middleware::Throttle::Lite::Backend::Abstract/"ABSTRACT METHODS">

=head2 reqs_done

See L<Plack::Middleware::Throttle::Lite::Backend::Abstract/"ABSTRACT METHODS">

=head2 increment

See L<Plack::Middleware::Throttle::Lite::Backend::Abstract/"ABSTRACT METHODS">

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/Wu-Wu/Plack-Middleware-Throttle-Lite-Backend-Memcached/issues>

=head1 SEE ALSO

L<Cache::Memcached::Fast>

L<Plack::Middleware::Throttle::Lite>

L<Plack::Middleware::Throttle::Lite::Backend::Abstract>

=head1 AUTHOR

Anton Gerasimov <chim@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Anton Gerasimov.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
